File: engine/appc/transform_store.py
"""Authoritative storage for every object's position and rotation.

One contract, two backends. The native backend (C++ contiguous array) is used
when _dauntless_host is present; the pure-Python backend is used headless.
They are never live at the same time, so this is one contract with two
implementations, not two sources of truth —
tests/unit/test_transform_store_conformance.py runs the identical suite
against both, which is what stops them drifting.

Handles are (index, generation). The generation counter exists so a stale
handle from a freed object fails loudly instead of silently reading whatever
object recycled its index.

Rotations are row-major nine-float tuples throughout, matching TGMatrix3's
storage and its mIJ = row I, column J naming. Column-vector convention: column
1 is forward (CLAUDE.md ↦ "Rotation matrix convention").
"""

_IDENTITY = (1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0)

# Twelve floats per slot: three position, nine rotation (row-major).
_STRIDE = 12


class StaleHandleError(RuntimeError):
    """A handle whose generation no longer matches its slot."""


class PythonTransformStore:
    """Flat-list backend. Used when the native extension is absent."""

    __slots__ = ("_data", "_generations", "_free", "_live")

    def __init__(self):
        self._data: list[float] = []
        self._generations: list[int] = []
        self._free: list[int] = []
        self._live: int = 0

    # ── Lifecycle ─────────────────────────────────────────────────────────────

    def alloc(self) -> tuple[int, int]:
        if self._free:
            index = self._free.pop()
            self._generations[index] += 1
        else:
            index = len(self._generations)
            self._generations.append(1)
            self._data.extend((0.0,) * _STRIDE)
        base = index * _STRIDE
        self._data[base:base + 3] = [0.0, 0.0, 0.0]
        self._data[base + 3:base + _STRIDE] = list(_IDENTITY)
        self._live += 1
        return index, self._generations[index]

    def free(self, index: int, generation: int) -> None:
        self._check(index, generation)
        # Bump on free as well as on reuse, so a handle to a freed-but-not-yet
        # reused slot is stale immediately rather than only after realloc.
        self._generations[index] += 1
        self._free.append(index)
        self._live -= 1

    def live_count(self) -> int:
        return self._live

    def capacity(self) -> int:
        return len(self._generations)

    # ── Access ────────────────────────────────────────────────────────────────

    def get_position(self, index: int, generation: int) -> tuple:
        self._check(index, generation)
        base = index * _STRIDE
        d = self._data
        return (d[base], d[base + 1], d[base + 2])

    def set_position(self, index: int, generation: int,
                     x: float, y: float, z: float) -> None:
        self._check(index, generation)
        base = index * _STRIDE
        d = self._data
        d[base] = float(x)
        d[base + 1] = float(y)
        d[base + 2] = float(z)

    def get_rotation(self, index: int, generation: int) -> tuple:
        self._check(index, generation)
        base = index * _STRIDE + 3
        return tuple(self._data[base:base + 9])

    def set_rotation(self, index: int, generation: int, t9) -> None:
        self._check(index, generation)
        base = index * _STRIDE + 3
        self._data[base:base + 9] = [float(v) for v in t9]

    def get_rotation_col(self, index: int, generation: int, col: int) -> tuple:
        self._check(index, generation)
        if col < 0 or col > 2:
            raise IndexError(col)
        base = index * _STRIDE + 3
        d = self._data
        return (d[base + col], d[base + 3 + col], d[base + 6 + col])

    # ── Internal ──────────────────────────────────────────────────────────────

    def _check(self, index: int, generation: int) -> None:
        gens = self._generations
        if index < 0 or index >= len(gens) or gens[index] != generation:
            raise StaleHandleError(
                "transform handle (%r, %r) is stale" % (index, generation))


class NativeTransformStore:
    """C++ backend. Thin translation of RuntimeError to StaleHandleError so
    both backends raise the same exception type."""

    __slots__ = ("_h",)

    def __init__(self, host_module):
        self._h = host_module

    def alloc(self) -> tuple[int, int]:
        return self._h.transform_alloc()

    def free(self, index: int, generation: int) -> None:
        try:
            self._h.transform_free(index, generation)
        except RuntimeError as exc:
            raise StaleHandleError(str(exc)) from exc

    def live_count(self) -> int:
        return self._h.transform_live_count()

    def capacity(self) -> int:
        return self._h.transform_capacity()

    def get_position(self, index: int, generation: int) -> tuple:
        try:
            return tuple(self._h.transform_get_position(index, generation))
        except RuntimeError as exc:
            raise StaleHandleError(str(exc)) from exc

    def set_position(self, index: int, generation: int,
                     x: float, y: float, z: float) -> None:
        try:
            self._h.transform_set_position(index, generation,
                                           float(x), float(y), float(z))
        except RuntimeError as exc:
            raise StaleHandleError(str(exc)) from exc

    def get_rotation(self, index: int, generation: int) -> tuple:
        try:
            return tuple(self._h.transform_get_rotation(index, generation))
        except RuntimeError as exc:
            raise StaleHandleError(str(exc)) from exc

    def set_rotation(self, index: int, generation: int, t9) -> None:
        try:
            self._h.transform_set_rotation(index, generation,
                                           [float(v) for v in t9])
        except RuntimeError as exc:
            raise StaleHandleError(str(exc)) from exc

    def get_rotation_col(self, index: int, generation: int, col: int) -> tuple:
        if col < 0 or col > 2:
            raise IndexError(col)
        try:
            return tuple(
                self._h.transform_get_rotation_col(index, generation, col))
        except RuntimeError as exc:
            raise StaleHandleError(str(exc)) from exc


_STORE = None


def get_store():
    """The process-wide transform store. Backend selected once, on first use.

    Resolved at USE, never at import: a module-level constant would freeze the
    choice before the host has had a chance to load the extension.
    """
    global _STORE
    if _STORE is None:
        try:
            import _dauntless_host as _h
        except ImportError:
            _h = None
        if _h is not None and hasattr(_h, "transform_alloc"):
            _STORE = NativeTransformStore(_h)
        else:
            _STORE = PythonTransformStore()
    return _STORE


def _reset_store_for_tests() -> None:
    """Drop the singleton so a test can start from an empty store."""
    global _STORE
    _STORE = None

File: native/src/transforms/src/transform_store.cc
#include "dauntless/transform_store.h"

#include <string>

namespace dauntless {

namespace {
constexpr float kIdentity[9] = {1.0f, 0.0f, 0.0f,
                                0.0f, 1.0f, 0.0f,
                                0.0f, 0.0f, 1.0f};
}  // namespace

StaleHandle::StaleHandle(std::uint32_t index, std::uint32_t generation)
    : std::runtime_error("transform handle (" + std::to_string(index) + ", " +
                         std::to_string(generation) + ") is stale") {}

std::pair<std::uint32_t, std::uint32_t> TransformStore::alloc() {
    std::uint32_t index;
    if (!free_.empty()) {
        index = free_.back();
        free_.pop_back();
        ++generations_[index];
    } else {
        index = static_cast<std::uint32_t>(generations_.size());
        generations_.push_back(1);
        slots_.emplace_back();
    }
    Transform& t = slots_[index];
    t.pos[0] = t.pos[1] = t.pos[2] = 0.0f;
    for (int i = 0; i < 9; ++i) t.rot[i] = kIdentity[i];
    ++live_;
    return {index, generations_[index]};
}

void TransformStore::free(std::uint32_t index, std::uint32_t generation) {
    check(index, generation);
    // Bump on free as well as on reuse, so a handle to a freed-but-not-yet
    // reused slot is stale immediately.
    ++generations_[index];
    free_.push_back(index);
    --live_;
}

bool TransformStore::valid(std::uint32_t index,
                           std::uint32_t generation) const {
    return index < generations_.size() && generations_[index] == generation;
}

void TransformStore::check(std::uint32_t index,
                           std::uint32_t generation) const {
    if (!valid(index, generation)) throw StaleHandle(index, generation);
}

std::array<float, 3> TransformStore::position(
        std::uint32_t index, std::uint32_t generation) const {
    check(index, generation);
    const Transform& t = slots_[index];
    return {t.pos[0], t.pos[1], t.pos[2]};
}

void TransformStore::set_position(std::uint32_t index, std::uint32_t generation,
                                  float x, float y, float z) {
    check(index, generation);
    Transform& t = slots_[index];
    t.pos[0] = x;
    t.pos[1] = y;
    t.pos[2] = z;
}

std::array<float, 9> TransformStore::rotation(
        std::uint32_t index, std::uint32_t generation) const {
    check(index, generation);
    const Transform& t = slots_[index];
    std::array<float, 9> out{};
    for (int i = 0; i < 9; ++i) out[i] = t.rot[i];
    return out;
}

void TransformStore::set_rotation(std::uint32_t index, std::uint32_t generation,
                                  const std::array<float, 9>& r) {
    check(index, generation);
    Transform& t = slots_[index];
    for (int i = 0; i < 9; ++i) t.rot[i] = r[i];
}

std::array<float, 3> TransformStore::rotation_col(
        std::uint32_t index, std::uint32_t generation, int col) const {
    check(index, generation);
    if (col < 0 || col > 2) throw std::out_of_range("rotation column");
    const Transform& t = slots_[index];
    return {t.rot[col], t.rot[3 + col], t.rot[6 + col]};
}

TransformStore& transform_store() {
    static TransformStore store;
    return store;
}

}  // namespace dauntless
